### PR TITLE
Frogger_OOP

### PR DESCRIPTION
**Pull request by Eugene50**
###Object Oriented JS Frogger

[Demo](https://eugene50.github.io/Frogger_OOP/) |
[Code base](https://github.com/Eugene50/Frogger_OOP)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/Eugene50/js_oop_frogger/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/.vscode/launch.json]: It is highly recommended not to have IDE specific files tracked by git.
Add a `.gitignore` file to your every project.
You may want to add a global `.gitignore`. Check this [helper](https://github.com/OleksiyRudenko/dev-env#global-git-helpers), the part re global `.gitignore`

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
+var Enemy = function() {
+    
+    this.sprite = 'images/enemy-bug.png';
+    this.x = 0;
+    this.y = 63;
+    this.speed = (function (min, max) {
+        return Math.floor(Math.random() * (max - min)) + min;
+      })(70,400);   //random speed of object   
+    this.width = 101;
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Hardcoded numbers are called "magic variables". Please, define and use constants.
Here and everywhere else across the code.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    this.x += (this.speed * dt);
+    } else { 
+        this.x = 0;
+    }
+};
+
+Enemy.prototype.render = function() {  //отрисовка жуков
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y,this.width,this.height);
+    
+        var  XColl=false;
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Please, fix code formatting. It is important as the code is written for the other people firsthand.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        var  YColl=false;
+              
+        if (((this.x+10) + (this.width-50) >= (player.x+10)) && ((this.x+10) <= (player.x+10) + (player.width-50))) XColl = true;
+        if (((this.y+20) + (this.height-120) >= player.y+10) && ((this.y+10) <= (player.y+10)+ player.height-120)) YColl = true;    
+             
+        if (XColl&&YColl){
+            player.x = 200;
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Classes should not refer to global variables.
Exceptions are global constants and library functions and variables.
Just pass player instance to the enemy at enemy creation, save it as a property and refer to it via class property.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+           this.y = 380;
+       };      
+    };
+
+Player.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y,this.width,this.height);    
+};
+
+Player.prototype.handleInput = function (key) {
+    if (key == 'left'&& this.x > 0){
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Please, tune up and use linters and code prettifiers in your IDE. These tools take care of code formatting, making code better readable.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    } else if (key == 'down'&& this.y < 380) {
+        this.y += 80;
+    }
+};
+
+
+var enemy = new Enemy();
+
+var enemy_1 = new Enemy();
+    enemy_1.y = 145;
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Initialize enemy position via constructor params.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    enemy_2.y = 228;
+
+var enemy_3 = new Enemy(); 
+var enemy_4 = new Enemy();
+    enemy_4.y = 145;   
+    
+        
+
+var allEnemies = [];
+    allEnemies.push( enemy,enemy_1,enemy_2,enemy_3,enemy_4);
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Just list enemies within `[]`
You can even instantiate those right within array definition. E.g.
```
var allEnemies = [
  new Enemy(63),
  new Enemy(145),
  new Enemy(228),
  // etc
];
```

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+
+var enemy = new Enemy();
+
+var enemy_1 = new Enemy();
+    enemy_1.y = 145;
+    
+
+var enemy_2 = new Enemy();
+    enemy_2.y = 228;
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: By convention variable names in JS are camelCased, using no underscores within the var names.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+document.addEventListener('keyup', function(e) {
+    var allowedKeys = {
+        37: 'left',
+        38: 'up',
+        39: 'right',
+        40: 'down'
+    };
+
+    player.handleInput(allowedKeys[e.keyCode]);
+});
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Add a newline.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        if (((this.y+20) + (this.height-120) >= player.y+10) && ((this.y+10) <= (player.y+10)+ player.height-120)) YColl = true;    
+             
+        if (XColl&&YColl){
+            player.x = 200;
+            player.y = 380;
+        };
+};     
+
+var Player = function () {
+    this.sprite = 'images/char-boy.png';
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Things like filepath are also good candidates to be constants.

**Comment by OleksiyRudenko** [comment]
The following self-check list may also help you to improve your code.

Minimal requirements to meet:
 * [ ] it is OK to employ ES6 features like `const`, `let` etc.
 * [ ] OO is implemented using JS prototype chain object model (**not** ES6 classes syntax)
 * Requirements re **Constants**:
   * [ ] all numbers like block dimensions, initial locations are defined as constants
   * [ ] there are core contants and derived constants
         (e.g. `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
   * [ ] arrays of constants are also constants 
         (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
   * [ ] const objects help organizing and structure const data even better
         (e.g. `const PLAYER_CONF = { initialPosition: {x: 1, y: 5}, sprite: '...', ...etc... };`
 * Requirements re **OOP**:
   * [ ] properties common for some classes are generalized into a base class
         (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
   * [ ] class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
         not `Subclass.prototype = new Superclass(params);`;
         [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
   * [ ] classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
         (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
         Hint: pass `Player` instance as an argument to every enemy
 * Most common mistakes
   * [ ] Make sure `target = condition ? valueWhenConditionTrue : valueWhenConditionFalse` is used instead of
         `condition ? target = valueWhenConditionTrue : target = valueWhenConditionFalse`;
         [Conditional (ternary) operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)
         
Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by OleksiyRudenko** [comment]
@Eugene50 Did you finish?
There several methods to let a code reviewer know you have finished an iteration:
- resolve conversations you believe you have properly processed
- let know explicitly
- respond to questions and inquiries (like  "Please, confirm all items above are positive" above)

Communication is a key.

File: submissions/Eugene50/js_oop_frogger/js/app.js (hunk 2 above)
Comment (Eugene50) [reply submissions/Eugene50/js_oop_frogger/js/app.js]: i made a function for this peace of code

**Comment by Eugene50** [comment]
please checkout my updates!)

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      this.x = 0;
+    }
+  };
+
+  Enemy.prototype.render = function () {
+    let XColl = false;
+    let YColl = false;
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y, this.width, this.height);
+
+    if (((this.x + 10) + (this.width - 50) >= (player.x + 10)) && ((this.x + 10) <= (player.x + 10) + (player.width - 50))) XColl = true;
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Here you still refer to `player`, which is a global variable.
Please, find a way to make an Enemy know the player explicitly, e.g. passing the `player` instance at construction time.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+      this.x = x;
+      this.y = y;
+      this.width = width;
+      this.height = height;
+      this.sprite = sprite;
+    }
+  }
+
+  Enemy.prototype.update = function (dt) {
+    if (this.x < 505) {
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: Hardcoded numbers are called "magic variables". Please, define and use constants.
Here and **everywhere else across the code**.

File: submissions/Eugene50/js_oop_frogger/js/app.js [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+      this.y -= 80;
+    } else if (key === 'down' && this.y < 380) {
+      this.y += 80;
+    }
+  };
+
+  const player = new Player(200, 380, 101, 171);
+
+  class Enemy {
+    constructor(speed, x, y, width, height, sprite = imgBug, boy = player) {
```
Comment (OleksiyRudenko) [inline submissions/Eugene50/js_oop_frogger/js/app.js]: There was a comment on that classes should not know anything about globals except for constants and library functions. Constructor params is a part of class.
You have introduced a param to pass an instance into class. You have to specify passed instance explicitly at construction time.
You may also remember a comment on default param values order from an earlier task.
Please, fix.

**Comment by OleksiyRudenko** [comment]
@Eugene50 Do you need any support?

**Comment by Eugene50** [comment]
Happy New Year!!!
i am lost 3 days because i`m back from Japan
And i realy have a little confused how to implement this thing..

**Comment by OleksiyRudenko** [comment]
@Eugene50 this way you will create a new player per each enemy. You already have the global `player`. Just pass it to every enemy at enemy construction. Not in constructor, but in the place where particular enemy construction happens.

**Comment by Eugene50** [comment]
Please checkout updates!

**Pull request merged**